examples/all-clusters-app/all-clusters-common/src/EnergyEvseDelegateImpl.cpp

```diff
@@ -0,0 +1,771 @@
+
+    // TODO: Generate events
+
+    return this->ComputeMaxChargeCurrentLimit();
+}
+
+/**
+ * @brief   Called when EVSE cluster receives EnableDischarging command
+ *
+ * @param dischargingEnabledUntil (can be null to indefinite discharging)
+ * @param maximumChargeCurrent (in mA)
+ */
+Status EnergyEvseDelegate::EnableDischarging(const DataModel::Nullable<uint32_t> & dischargingEnabledUntil,
+                                             const int64_t & maximumDischargeCurrent)
+{
+    ChipLogProgress(AppServer, "EnergyEvseDelegate::EnableDischarging() called.");
+
+    /* update SupplyState */
+    SetSupplyState(SupplyStateEnum::kDischargingEnabled);
+
+    // TODO: Generate events
+
+    return Status::Success;
+}
+
+/**
+ * @brief    Called when EVSE cluster receives StartDiagnostics command
+ */
+Status EnergyEvseDelegate::StartDiagnostics()
+{
+    /* For EVSE manufacturers to customize */
+    ChipLogProgress(AppServer, "EnergyEvseDelegate::StartDiagnostics()");
+
+    /* update SupplyState */
+    SetSupplyState(SupplyStateEnum::kDisabledDiagnostics);
+
+    // TODO: Generate events
+
+    return Status::Success;
+}
+
+/* ---------------------------------------------------------------------------
+ * FUNCTIONS BELOW:
+ *    - EVSE Hardware interface
+ *
+ *  RegisterEvseHardwareCallback( callbackType, callbackFnc )
+ *  SetMaxHardwareCurrentLimit( currentmA )
+ *  SetCircuitCapacity( currentmA )
+ *  SetCableAssemblyLimit( currentmA )
+ *  SetState( EVSEStateEnum )
+ *  SetFault
+ *
+ */
+
+/**
+ * @brief    Called by EVSE Hardware to register a callback
+ *
+ * @param    Callback function
+ */
+#if 0
+Status EnergyEvseDelegate::HwRegisterEvseHardwareCallback(int Callback) // TODO
+{
+    // TODO
+    return CHIP_NO_ERROR;
+}
+#endif
+/**
+ * @brief    Called by EVSE Hardware to notify the delegate of the maximum
+ *           current limit supported by the hardware.
+ *
+ *           This is normally called at start-up.
+ *
+ * @param    currentmA
+ */
+Status EnergyEvseDelegate::HwSetMaxHardwareCurrentLimit(int64_t currentmA)
+{
+    if (currentmA < kMinimumChargeCurrent || currentmA > kMaximumChargeCurrent)
+    {
+        return Status::ConstraintError;
+    }
+
+    /* there is no attribute to store this so store in private variable */
+    mMaxHardwareCurrentLimit = currentmA;
+
+    return this->ComputeMaxChargeCurrentLimit();
+}
+
+/**
+ * @brief    Called by EVSE Hardware to notify the delegate of maximum electrician
+ *           set current limit.
+ *
+ *           This is normally called at start-up when reading from DIP-switch
+ *           settings.
+ *
+ * @param    currentmA
+ */
+Status EnergyEvseDelegate::HwSetCircuitCapacity(int64_t currentmA)
+{
+    if (currentmA < kMinimumChargeCurrent || currentmA > kMaximumChargeCurrent)
+    {
+        return Status::ConstraintError;
+    }
+
+    mCircuitCapacity = currentmA;
+    MatterReportingAttributeChangeCallback(mEndpointId, EnergyEvse::Id, CircuitCapacity::Id);
```

Comment:
This is done here, but not in the other `Set` above. Why?

Reply:
same question for below.

Reply:
As the comment says - there are a couple of internal HW based current limits, but they don't have (or need) attributes.


    /* there is no attribute to store this so store in private variable */
    mMaxHardwareCurrentLimit = currentmA;
    
    return this->ComputeMaxChargeCurrentLimit();

The ComputeMaxChargeCurrentLimit works out the minimum of several current limits and that becomes the MaximumChargeCurrentLimit and in that function it calls MatterReportingAttributeChangeCallback if it has changed.